### PR TITLE
Product quantization

### PR DESCRIPTION
**Pull request by IvanPleshkov**
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [ ] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [ ] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by timvisee — APPROVED** [review]
LGTM. I like this generic approach for the scorer much better than what we had before. Thanks for talking me through the main bits in a call.

I wonder if we could simplify the `crate_scalar` and `create_pq` methods a bit, but I guess it is fine as is.

Andrey should probably take a look at the public interface/API design.

File: lib/segment/src/vector_storage/memmap_vector_storage.rs [hunk 1]
```diff
@@ -603,7 +603,7 @@ mod tests {
         }
         .into();
 
-        borrowed_storage.quantize(dir.path(), &config).unwrap();
+        borrowed_storage.quantize(dir.path(), &config, 1).unwrap();
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/memmap_vector_storage.rs:606]: That's scalar quantization test. In scalar quantization there are no any parallel encoding (not requires, it's fast enough)

File: lib/segment/src/vector_storage/quantized/quantized_raw_scorer.rs [hunk 2]
```diff
@@ -1,33 +1,24 @@
...
 {
-    query: TEncodedQuery,
+    pub(crate) query: TEncodedQuery,
     /// [`BitSlice`] defining flags for deleted points (and thus these vectors).
-    point_deleted: &'a BitSlice,
+    pub(crate) point_deleted: &'a BitSlice,
     /// [`BitSlice`] defining flags for deleted vectors in this segment.
-    vec_deleted: &'a BitSlice,
+    pub(crate) vec_deleted: &'a BitSlice,
     // Total number of vectors including deleted ones
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/quantized/quantized_raw_scorer.rs:16]: Perfect! Good catch! I missed it while resolving conflicts

File: lib/segment/src/vector_storage/quantized/quantized_raw_scorer.rs [hunk 3]
```diff
@@ -1,33 +1,24 @@
...
     TEncodedVectors: quantization::EncodedVectors<TEncodedQuery>,
 {
-    query: TEncodedQuery,
+    pub(crate) query: TEncodedQuery,
     /// [`BitSlice`] defining flags for deleted points (and thus these vectors).
-    point_deleted: &'a BitSlice,
+    pub(crate) point_deleted: &'a BitSlice,
     /// [`BitSlice`] defining flags for deleted vectors in this segment.
-    vec_deleted: &'a BitSlice,
+    pub(crate) vec_deleted: &'a BitSlice,
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/quantized/quantized_raw_scorer.rs]: thanks, done

File: lib/segment/src/vector_storage/quantized/quantized_vectors.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        };
+        Ok(())
+    }
+
+    pub fn files(&self) -> Vec<PathBuf> {
+        let mut result = vec![self.path.join(QUANTIZED_CONFIG_PATH)];
+        let storage_files: Vec<PathBuf> =
+            vec![QUANTIZED_DATA_PATH.into(), QUANTIZED_META_PATH.into()];
+        result.extend(storage_files.into_iter().map(|file| self.path.join(file)));
+        result
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/quantized/quantized_vectors.rs]: I guess you were surprised while suggesting it =) It's really overcomplicated

File: lib/segment/src/vector_storage/quantized/quantized_vectors.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+            path: path.to_path_buf(),
+            distance,
+        };
+
+        quantized_vectors.save_to(path)?;
+        atomic_save_json(&path.join(QUANTIZED_CONFIG_PATH), &quantized_vectors.config)?;
+        Ok(quantized_vectors)
+    }
+
+    pub fn check_exists(path: &Path) -> bool {
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/quantized/quantized_vectors.rs]: Agreed, I'm bad at namings. I changed it to `config_exists`, thanks for the variant

[the remark this answers is not in the document]
Comment (timvisee) [reply lib/segment/src/vector_storage/quantized/quantized_vectors.rs]: Naming is the hardest thing in programming.

**Pull request merged**